File: 3967-earliest-finish-time-for-land-and-water-rides-ii/earliest-finish-time-for-land-and-water-rides-ii.cs
public class Solution {
    public int EarliestFinishTime(int[] landStartTime, int[] landDuration, int[] waterStartTime, int[] waterDuration) {
        int minL, minW, res, m=landStartTime.Length, n=waterStartTime.Length;
        minL=minW=res=100000000;
        for(int i=0; i<m; i++){
            minL=Math.Min(minL, landStartTime[i]+landDuration[i]);
        }
        for(int i=0; i<n; i++){
            minW=Math.Min(minW, waterStartTime[i]+waterDuration[i]);
            res=Math.Min(res, Math.Max(minL, waterStartTime[i])+waterDuration[i]);
        }
        for(int i=0; i<m; i++){
            res=Math.Min(res, Math.Max(minW, landStartTime[i])+landDuration[i]);
        }
        return res;
    }
}

File: 3967-earliest-finish-time-for-land-and-water-rides-ii/earliest-finish-time-for-land-and-water-rides-ii.py
class Solution:
    def earliestFinishTime(self, landStartTime: List[int], landDuration: List[int], waterStartTime: List[int], waterDuration: List[int]) -> int:
        minL=minW=res=float('inf')
        m=len(landStartTime);n=len(waterStartTime)
        for i in range(m):
            minL=min(minL, landStartTime[i]+landDuration[i])
        for i in range(n):
            minW=min(minW, waterStartTime[i]+waterDuration[i])
            res=min(res, max(minL, waterStartTime[i])+waterDuration[i])
        for i in range(m):
            res=min(res, max(minW, landStartTime[i])+landDuration[i])
        return res